[Merged by Bors] - Add `validator-manager` (#3502)
sigp/lighthouse: book/src/validator-manager-move.md

# Moving Validators

The `lighthouse validator-manager move` command uses the VC HTTP API to move
validators from one VC (the "src" VC) to another VC (the "dest" VC). The move
operation is *comprehensive*; it will:

> - Disable the validators the src VC.

Reviewer: ```suggestion
- Disable the validators on the src VC.
```

dest VC. Therefore, it is recommended to **never use the `move` command without
having a backup of all validator keystores (e.g. the mnemonic).**

## Simple Example

The following command will move all validators from the VC running at
`http://localhost:6062` to the VC running at `http://localhost:5062`.

```bash
lighthouse \
	validator-manager \
	move \
	--src-vc-url http://localhost:6062 \
	--src-vc-token ~/src-token.txt \
	--dest-vc-url http://localhost:5062 \
	--dest-vc-token ~/.lighthouse/mainnet/validators/api-token.txt \
	--validators all
```

## Detailed Guide

This guide describes the steps to move validators between two validator clients (VCs) which are
able to SSH between each other. This guide assumes experience with the Linux command line and SSH
connections.

There will be two VCs in this example:

- The *source* VC which contains the validators/keystores to be moved.
- The *destination* VC which is to take the validators/keystores from the source.

There will be two hosts in this example:

- Host 1 (*"source host"*): Is running the `src-vc`.
- Host 2 (*"destination host"*): Is running the `dest-vc`.

> The example will assume that all commands are run on Host 1. It also assumes

Reviewer 2: This can be confusing since Steps 2 and 5 are run on Host 2 (destination host), hence the suggestion to change, and also added info in Step 4.

```suggestion
The example assumes
```

that Host 1 is able to SSH to Host 2.

In reality, many host configurations are possible. For example:

- Both VCs on the same host.
- Both VCs on different hosts and the `validator-manager` being used on a third host.

### 1. Configure the Source VC

> The source VC needs to have the following flags at a mininum:

Reviewer 2: ```suggestion
The source VC needs to have the following flags at a minimum:
```

- `--http`
- `--unencrypted-http-transport`
- `--http-address 127.0.0.1`
- `--http-port 5062`
- `--http-allow-keystore-export`

Therefore, the source VC command might look like:

```bash
lighthouse \
    vc \
    --http \
    --unencrypted-http-transport \
    --http-address 127.0.0.1 \
    --http-port 5062 \
    --http-allow-keystore-export
```

### 2. Configure the Destination VC

> The destination VC needs to have the following flags at a mininum:

Reviewer 2: ```suggestion
The destination VC needs to have the following flags at a minimum:
```

- `--http`
- `--unencrypted-http-transport`
- `--http-address 127.0.0.1`
- `--http-port 5062`
- `--enable-doppelganger-protection`

Therefore, the destination VC command might look like:

```bash
lighthouse \
    vc \
    --http \
    --unencrypted-http-transport \
    --http-address 127.0.0.1 \
    --http-port 5062 \
    --enable-doppelganger-protection
```

> The `--enable-doppelganger-protection` flag is not *strictly* required, however

> > it is recommended for an additional layer of safety. It will result in 3-4

Reviewer 2: Same comment as in `validator-manager-create.md`
```suggestion
> it is recommended for an additional layer of safety. It will result in 2-3
```

> epochs of downtime for the validator after it is moved, which is generally an
> inconsequential cost in lost rewards or penalties.
> 
> Optionally, users can add the `--http-store-passwords-in-secrets-dir` flag if they'd like to have
> the import validator keystore passwords stored in separate files rather than in the

> > `valdiator-definitions.yml` file. If you don't know what this means, you can safely omit the flag.

Reviewer 2: ```suggestion
> `validator-definitions.yml` file. If you don't know what this means, you can safely omit the flag.
```

> 

Reviewer 2: Added the configurations required if both VCs are in the same host.
````suggestion
Note: If both VCs are on the same host, a different `http-port` value is required (e.g., `--http-port 5063`). Therefore, the command will become:
```bash
lighthouse \
    vc \
    --http \
    --unencrypted-http-transport \
    --http-address 127.0.0.1 \
    --http-port 5063 \
    --enable-doppelganger-protection
```
````

### 3. Obtain the Source API Token

The VC API is protected by an *API token*. This is stored in a file on each of the hosts. Since
we'll be running our command on the destination host, it will need to have the API token for the
source host on its file-system.

On the **source host**, find the location of the `api-token.txt` file and copy the contents. The
location of the file varies, but it is located in the "validator directory" of your data directory,

> alongside validator keystores. For example: `~/.lighthouse/mainnet/validators/api-token.txt`.

Reviewer 2: ```suggestion
alongside validator keystores. For example: `~/.lighthouse/mainnet/validators/api-token.txt`. If you are unsure of the `api-token.txt` path, you can run `curl http://localhost:5062/lighthouse/auth` which will show the path.
```

Copy the contents of that file into a new file on the **destination host** at `~/src-token.txt`. The
API token should be similar to `api-token-0x03eace4c98e8f77477bb99efb74f9af10d800bd3318f92c33b719a4644254d4123`.

> 

Reviewer 2: Added the configurations required if both VCs are in the same host.
```suggestion
If both VCs are on the same host, you can skip copying the `api-token.txt` to the destination host, since now both `api-token.txt` files reside on the same host. You only have to identify the directory of the `api-token.txt` of both VCs. If you are unsure of the `api-token.txt` path, you can run (assuming the `http-port` used are 5062 and 5063):
`curl http://localhost:5062/lighthouse/auth` 
`curl http://localhost:5063/lighthouse/auth` 
```

### 4. Create an SSH Tunnel

> In one terminal window, SSH to the **destination host** and establish a reverse-SSH connection

Reviewer 2: ```suggestion
In the **source host**, open a terminal window, SSH to the **destination host** and establish a reverse-SSH connection
```

> between the **desination host** and the **source host**.

Reviewer 2: ```suggestion
between the **destination host** and the **source host**.
```

```bash
ssh dest-host
ssh -L 6062:localhost:5062 src-host
```

It's important that you leave this session open throughout the rest of this tutorial. If you close
this terminal window then the connection between the destination and source host will be lost.

> 

Reviewer 2: Added the configurations required if both VCs are in the same host.
```suggestion
> You can skip this step if both VCs are on the same host. 
```

### 5. Move

With the SSH tunnel established between the `dest-host` and `src-host`, from the **destination
host** run the command to move the validators:

```bash
lighthouse \
	validator-manager \
	move \
	--src-vc-url http://localhost:6062 \
	--src-vc-token ~/src-token.txt \
	--dest-vc-url http://localhost:5062 \
	--dest-vc-token ~/.lighthouse/mainnet/validators/api-token.txt \
	--validators all
```

The command will provide information about the progress of the operation and
emit `Done.` when the operation has completed successfully. For example:

```bash

> Running validator manager for mainnet network

Reviewer 2: Same comment as with the `import` command
```suggestion
Running validator manager
```

Validator client is reachable at http://localhost:5062/ and reports 2 validators
Validator client is reachable at http://localhost:6062/ and reports 0 validators
Moved keystore 1 of 2
Moved keystore 2 of 2
Done.
```

> 

Reviewer 2: Added more information on the VC log:
````suggestion
At the same time, `lighthouse vc` will log:
```bash
INFO Importing keystores via standard HTTP API, count: 1
INFO Enabled validator                       voting_pubkey: 0xab6e29f1b98fedfca878edce2b471f1b5ee58ee4c3bd216201f98254ef6f6eac40a53d74c8b7da54f51d3e85cacae92f, signing_method: local_keystore
INFO Modified key_cache saved successfully
````

Once the operation completes successfully, there is nothing else to be done. The
validators have been removed from the `src-host` and enabled at the `dest-host`.

> If the `--enable-doppelganger-protection` flag was used it may take 3-4 epochs

Reviewer 2: ```suggestion
If the `--enable-doppelganger-protection` flag was used it may take 2-3 epochs
```

for the validators to start attesting and producing blocks on the `dest-host`.

> 

Reviewer 2: Add the scenario with only moving some of the validators:
````suggestion
If you would only like to move some validators, you can replace the flag `--validators all` with one or more validator public keys. For example:

```bash
lighthouse \
	validator-manager \
	move \
	--src-vc-url http://localhost:6062 \
	--src-vc-token ~/src-token.txt \
	--dest-vc-url http://localhost:5062 \
	--dest-vc-token ~/.lighthouse/mainnet/validators/api-token.txt \
	--validators 0x9096aab771e44da149bd7c9926d6f7bb96ef465c0eeb4918be5178cd23a1deb4aec232c61d85ff329b54ed4a3bdfff3a,0x90fc4f72d898a8f01ab71242e36f4545aaf87e3887be81632bb8ba4b2ae8fb70753a62f866344d7905e9a07f5a9cdda1
```
````

Reviewer 2: Added the configurations required if both VCs are in the same host.
````suggestion
> If both VCs are on the same host, the command would become:
```bash
lighthouse \
	validator-manager \
	move \
	--src-vc-url http://localhost:5062 \
	--src-vc-token ~/src-token.txt \
	--dest-vc-url http://localhost:5063 \
	--dest-vc-token ~/dest-token.txt \
	--validators all
```
````

> Any errors encounted during the operation should include information on how to

Reviewer 2: ```suggestion
Any errors encountered during the operation should include information on how to
```
